sw/device/lib/runtime/otbn.c

```diff
@@ -63,15 +78,20 @@ otbn_result_t otbn_init(otbn_t *ctx, mmio_region_t base_addr) {
 }
 
 otbn_result_t otbn_load_app(otbn_t *ctx, const otbn_app_t app) {
-  if (app.imem_end <= app.imem_start || app.dmem_end < app.dmem_start) {
+  if (app.imem_end <= app.imem_start ||
+      app.dmem_data_end < app.dmem_data_start ||
+      app.dmem_bss_end < app.dmem_bss_start ||
+      (uintptr_t)app.dmem_bss_offset <
+          (uintptr_t)app.dmem_data_end - (uintptr_t)app.dmem_data_start) {
```

Comment:
I think these checks are:
- The IMEM range isn't backwards
- The DMEM data range isn't backwards
- The DMEM BSS range isn't backwards
- The BSS offset lies above the top of initialised DMEM.

which looks good to me.  I wonder whether it makes sense to define a "`check_wellformed_range`" helper function to do the first three checks. It would be kind of trivial and get inlined, but it would help to document what's going on. Alternatively, you could split this into 4 `if` statements with comments. But that's a bit sad because you have to repeat the return statement.

Also, if you change `dmem_bss_offset` to a `uint32_t` as I suggested in another comment then I think you can get rid of all the casts here: the pointers being subtracted are `uint8_t*`, so subtraction will give the byte difference.

The same comments apply on the `silicon_creator` side.